Automattic/wp-calypso
# WooCommerce: Update state structure

On the call today, we talked about refactoring state into a simpler structure, specifically renaming `wc-api` into `sites`, so the state tree would look more like:

```
state.extensions.woocommerce.sites[siteId].products
state.extensions.woocommerce.sites[siteId].settings
state.extensions.woocommerce.sites[siteId].settings.general
state.extensions.woocommerce.sites[siteId].paymentMethods
state.extensions.woocommerce.sites[siteId].shippingZones
```

I'm sorry, this is huge. There's no way around it - I've moved files from `state/site` and `state/wc-api` into `state/sites`. This also meant I had to update the paths, and changing the shape of the state means I also had to update the tests. In doing so, I've added action tests, and split out the selector tests into 1 describe per function (so these can be moved later).

I tried to keep each commit as one unit of changes, a move or rename, and the commit message should reflect that change. The commit that will need the most review is d73fa86a7b92bc40abee3055a007a2df6759d40d, which:

- reformats reducers: updating the `wc-api` to export `createReducer`'d reducers, rather than just a map of action-types/functions
- strips out the `payload` of actions, so actions are now flatter: type, siteId, and data are usually all they return
- adds action tests
- updates the selector tests as mentioned above
- updates the reducer tests now that they're reducers, not just mapped functions

This PR also 

- adds a `constants.js` file containing (currently only) `LOADING`, so that we don't have to redefine it in every reducer.
- changes the `settingsGeneral` subtree to `settings.general`, so we can add other settings subtrees (assuming we'll need this based on the API endpoints).

I didn't really touch the UI sub-tree other than fixing paths.

**To test**

- Run the tests: `npm run test-client client/extensions/woocommerce/`
- Check that the actions & state you're relying on in your components is still there
